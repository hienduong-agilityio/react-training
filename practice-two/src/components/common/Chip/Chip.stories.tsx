import type { Meta, StoryObj } from '@storybook/react';
import Chip, { CHIP_COLOR } from './index';

const meta = {
  title: 'Components/common/Chip',
  component: Chip,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs']
} satisfies Meta<typeof Chip>;

export default meta;

type Story = StoryObj<typeof meta>;

export const FireType: Story = {
  args: {
    label: 'fire',
    color: CHIP_COLOR.FIRE
  }
};

export const BugType: Story = {
  args: {
    label: 'bug',

    color: CHIP_COLOR.BUG
  }
};

export const WaterType: Story = {
  args: {
    label: 'water',
    color: CHIP_COLOR.WATER
  }
};

export const FlyingType: Story = {
  args: {
    label: 'flying',
    color: CHIP_COLOR.FLYING
  }
};

export const PoisonType: Story = {
  args: {
    label: 'poison',
    color: CHIP_COLOR.POISON
  }
};

export const RockType: Story = {
  args: {
    label: 'rock',
    color: CHIP_COLOR.ROCK
  }
};

export const GroundType: Story = {
  args: {
    label: 'ground',
    color: CHIP_COLOR.GROUND
  }
};

export const GhostType: Story = {
  args: {
    label: 'ghost',
    color: CHIP_COLOR.GHOST
  }
};

export const ElectricType: Story = {
  args: {
    label: 'electric',
    color: CHIP_COLOR.ELECTRIC
  }
};
