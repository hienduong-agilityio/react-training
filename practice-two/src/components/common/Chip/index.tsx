// Library
import classNames from 'classnames';
import { ReactNode } from 'react';

export interface IChipProps {
  color: CHIP_COLOR;
  icon?: ReactNode;
  label: string;
  customClasses?: string;
}

export enum CHIP_COLOR {
  FIRE = 'fire',
  BUG = 'bug',
  WATER = 'water',
  FLYING = 'flying',
  POISON = 'poison',
  NORMAL = 'normal',
  ROCK = 'rock',
  GROUND = 'ground',
  FIGHTING = 'fighting',
  GHOST = 'ghost',
  PSYCHIC = 'psychic',
  ICE = 'ice',
  DRAGON = 'dragon',
  DARK = 'dark',
  FAIRY = 'fairy',
  ELECTRIC = 'electric',
  STEEL = 'steel',
  GRASS = 'grass'
}

const chipDefault: string = 'inline-block rounded-full py-1 px-4 capitalize';

const colorClasses: { [key in CHIP_COLOR]: string } = {
  fire: 'bg-fire',
  bug: 'bg-bug',
  water: 'bg-water',
  flying: 'bg-flying',
  poison: 'bg-poison',
  normal: 'bg-normal',
  rock: 'bg-rock',
  ground: 'bg-ground',
  fighting: 'bg-fighting',
  ghost: 'bg-ghost',
  psychic: 'bg-psychic',
  ice: 'bg-ice',
  dragon: 'bg-dragon',
  dark: 'bg-dark',
  fairy: 'bg-fairy',
  electric: 'bg-electric',
  steel: 'bg-steel',
  grass: 'bg-grass'
};

/**
 * Chip Component
 * @param color - The background color of the chip.
 * @param icon - The icon to be displayed on the chip.
 * @param label - The content to be displayed inside the chip.
 *
 * @returns {JSX.Element} - The Chip component.
 */

const Chip = ({ color = CHIP_COLOR.NORMAL, icon, label = '', customClasses = '' }: IChipProps): JSX.Element => {
  const chipClasses: string = classNames([chipDefault, colorClasses[color], customClasses]);

  return (
    <div className={chipClasses}>
      {icon && <span className="chip-icon">{icon}</span>}
      <span className="chip-label">{label}</span>
    </div>
  );
};

export default Chip;
