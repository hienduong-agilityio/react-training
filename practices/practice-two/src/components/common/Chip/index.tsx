// Library
import classNames from 'classnames';
import { ReactNode } from 'react';

export interface IChipProps {
  color?: CHIP_COLOR;
  icon?: ReactNode;
  label: string;
  customClasses?: string;
}

export enum CHIP_COLOR {
  FIRE = 'fire',
  BUG = 'bug',
  WATER = 'water',
  FLYING = 'flying',
  POISON = 'poison',
  NORMAL = 'normal',
  ROCK = 'rock',
  GROUND = 'ground',
  FIGHTING = 'fighting',
  GHOST = 'ghost',
  PSYCHIC = 'psychic',
  ICE = 'ice',
  DRAGON = 'dragon',
  DARK = 'dark',
  FAIRY = 'fairy',
  ELECTRIC = 'electric',
  STEEL = 'steel',
  GRASS = 'grass'
}

const chipBackground: { [key in CHIP_COLOR]: string } = {
  fire: 'bg-fire-300',
  bug: 'bg-bug-300',
  water: 'bg-water-300',
  flying: 'bg-flying-300',
  poison: 'bg-poison-300',
  normal: 'bg-normal-300',
  rock: 'bg-rock-300',
  ground: 'bg-ground-300',
  fighting: 'bg-fighting-300',
  ghost: 'bg-ghost-300',
  psychic: 'bg-psychic-300',
  ice: 'bg-ice-300',
  dragon: 'bg-dragon-300',
  dark: 'bg-dark-300',
  fairy: 'bg-fairy-300',
  electric: 'bg-electric-300',
  steel: 'bg-steel-300',
  grass: 'bg-grass-300'
};

const chipDefault: string = 'inline-block rounded-full py-1 px-4 capitalize';

/**
 * Chip Component
 * @param color - The background color of the chip.
 * @param icon - The icon to be displayed on the chip.
 * @param label - The content to be displayed inside the chip.
 *
 * @returns {JSX.Element} - The Chip component.
 */

const Chip = ({ color = CHIP_COLOR.NORMAL, icon, label = '', customClasses = '' }: IChipProps): JSX.Element => {
  const chipClasses: string = classNames([chipDefault, chipBackground[color], customClasses]);

  return (
    <div className={chipClasses}>
      {icon && <span className="chip-icon">{icon}</span>}
      <span className="chip-label">{label}</span>
    </div>
  );
};

export default Chip;
