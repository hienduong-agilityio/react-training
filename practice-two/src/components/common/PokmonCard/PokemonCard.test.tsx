import { render } from '@testing-library/react';
import PokemonCard from './index';
import { CHIP_COLOR } from '../Chip';
import '@testing-library/jest-dom';

describe('PokemonCard component', () => {
  const mockProps = {
    pokemonID: '001',
    pokemonName: 'Bulbasaur',
    pokemonType: [CHIP_COLOR.GRASS, CHIP_COLOR.POISON],
    pokemonImg: 'https://example.com/bulbasaur.png'
  };

  it('renders correctly with given props', () => {
    const { asFragment, getByText, getByAltText } = render(<PokemonCard {...mockProps} />);

    // Check if the name is rendered
    expect(getByText('Bulbasaur')).toBeInTheDocument();

    // Check if the ID is rendered
    expect(getByText('#001')).toBeInTheDocument();

    // Check if the image is rendered
    expect(getByAltText('Bulbasaur')).toBeInTheDocument();

    // Check if the chips for types are rendered
    expect(getByText('grass')).toBeInTheDocument();
    expect(getByText('poison')).toBeInTheDocument();

    // Update snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
