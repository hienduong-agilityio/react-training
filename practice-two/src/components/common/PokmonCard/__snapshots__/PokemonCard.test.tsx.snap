// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard component renders correctly with given props 1`] = `
<DocumentFragment>
  <div
    class="w-full h-full flex flex-col items-center rounded-lg pt-16 relative"
  >
    <div
      class="w-full h-full rounded-xl absolute top-0 left-0 overflow-hidden z-0 backface-visibility-hidden"
    >
      <div
        class="w-full h-full absolute top-0 left-1/2 transform -translate-x-1/2 transition duration-500 bg-no-repeat bg-right-top	 bg-pokemon-card bg-grass-400"
      />
    </div>
    <div
      class="absolute -top-12 transform -translate-y-10"
    >
      <img
        alt="The pokemon images"
        class="w-40 h-40"
        src="https://example.com/bulbasaur.png"
      />
    </div>
    <div
      class="z-10 w-full p-5 px-5"
    >
      <span
        class="text-2xl font-bold capitalize block my-2"
      >
        Bulbasaur
      </span>
      <p
        class="text-black text-base"
      >
        #001
      </p>
      <div
        class="inline-block rounded-full py-1 px-4 capitalize bg-grass-300 mr-2"
      >
        <span
          class="chip-label"
        >
          grass
        </span>
      </div>
      <div
        class="inline-block rounded-full py-1 px-4 capitalize bg-poison-300 mr-2"
      >
        <span
          class="chip-label"
        >
          poison
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
